File: datastruct-algoritem-java/src/main/java/datastruct/tree/AvlTree.java
package datastruct.tree;

/**
 * AVL树：AVL树是平衡搜索树，空树是AVL树，AVL树定义：
 * 1.左子树和右子树是AVL树
 * 2.左子树的高度-右子树的高度<=1
 * 3.它是搜索树
 * 相比于搜索树，多了一个高度和平衡因子函数进行旋转
 *
 * @author czx
 * @date 2019/12/26
 */
public class AvlTree<T extends Comparable<? super T>> {
    AvlNode root;

    public void insert(T t) {
        root = insert(root, t);
    }

    public void remove(T t) {
        root = remove(root, t);
    }

    public void inOrder() {
        inOrder(root);
    }

    public void inOrder(AvlNode node) {
        if (node == null) {
            return;
        }
        if (node.left != null) {
            inOrder(node.left);
        }
        System.out.print(node.value + " ");
        if (node.right != null) {
            inOrder(node.right);
        }
    }

    private AvlNode insert(AvlNode node, T t) {
        if (node == null) {
            return new AvlNode(t);
        }
        int comp = t.compareTo(node.value);
        if (comp < 0) {
            node.left = insert(node.left, t);
        } else if (comp > 0) {
            node.right = insert(node.right, t);
        }
        return balance(node);
    }

    private AvlNode remove(AvlNode node, T t) {
        if (node == null) {
            return null;
        }
        int compare = t.compareTo(node.value);

        if (compare < 0) {
            node.left = remove(node.left, t);
        } else if (compare > 0) {
            node.right = remove(node.right, t);
        } else if (node.left != null && node.right != null) {
            T value = findMin(node.right).value;
            node.right = remove(node.right, value);
            node.value = value;
        } else {
            node = node.left != null ? node.left : node.right;
        }
        return balance(node);
    }

    private AvlNode findMin(AvlNode node) {
        while (node.left != null) {
            node = node.left;
        }
        return node;
    }

    // 平衡树
    private AvlNode balance(AvlNode node) {
        if (node == null) {
            return null;
        }
        // 如果左子树大于右子树,说明插入的节点是在左子树
        if (height(node.left) - height(node.right) > 1) {
            // 如果左子树的左子树高度大于左子树的右子树高度，单旋转
            if (height(node.left.left) >= height(node.left.right)) {
                // 单旋转
                node = rotateWithLeft(node);
            } else {
                // 双旋转
                node = doubleWithLeft(node);
            }
        } else if (height(node.right) - height(node.left) > 1) {
            if (height(node.right.right) >= height(node.right.left)) {
                node = rotateWithRight(node);
            } else {
                node = doubleWithRight(node);
            }
        }
        node.h = Math.max(height(node.left), height(node.right)) + 1;
        return node;
    }

    // 左单旋
    private AvlNode rotateWithLeft(AvlNode k2) {
        AvlNode k1 = k2.left;
        k2.left = k1.right; // 当前节点左子树指向他的左节点的右子树
        k1.right = k2; // 当前节点的右子树指向当前节点
        k2.h = Math.max(height(k2.left), height(k2.right)) + 1;// 调整后的右子树的最大高度
        k1.h = Math.max(height(k1.left), k2.h) + 1; // 调整后的根节点的最大高度
        return k1;
    }

    // 右单旋
    private AvlNode rotateWithRight(AvlNode k1) {
        AvlNode k2 = k1.right;
        k1.right = k2.left;
        k2.left = k1;
        k1.h = Math.max(height(k1.left), height(k1.right)) + 1;
        k2.h = Math.max(height(k2.right), k1.h) + 1;
        return k2;
    }

    // 左双旋
    private AvlNode doubleWithLeft(AvlNode k3) {
        k3.left = rotateWithRight(k3.left);
        return rotateWithLeft(k3);
    }

    // 右双旋
    private AvlNode doubleWithRight(AvlNode k2) {
        k2.right = rotateWithLeft(k2.right);
        return rotateWithRight(k2);
    }

    private int height(AvlNode node) {
        return node == null ? -1 : node.h;
    }

    private class AvlNode {
        T value;
        AvlNode left;
        AvlNode right;
        int h;

        public AvlNode(T value) {
            this(value, null, null, 0);
        }

        public AvlNode(T value, AvlNode left, AvlNode right, int h) {
            this.value = value;
            this.left = left;
            this.right = right;
            this.h = h;
        }
    }
}

File: datastruxt-algoritem-go/datastruct/tree/avltree/avltree.go
package avltree

import (
	"fmt"
	"math"
)

// Create by czx on 2019/12/26

type node struct {
	value interface{}
	left  *node
	right *node
	h     int
}

func newNode(value interface{}, left *node, right *node, h int) *node {
	return &node{value: value, left: left, right: right, h: h}
}

type AvlTree struct {
	root *node
	fn   func(c1, c2 interface{}) int
}

func NewAvlTree(fn func(c1, c2 interface{}) int) *AvlTree {
	return &AvlTree{fn: fn}
}

func (at *AvlTree) Insert(value interface{}) {
	at.root = at.insert(at.root, value)
}

func (at *AvlTree) Remove(value interface{}) {
	at.root = at.remove(at.root, value)
}

func (at *AvlTree) InOrder() {
	at.inOrder(at.root)
	fmt.Println()
}

func (at *AvlTree) insert(n *node, value interface{}) *node {
	if n == nil {
		return newNode(value, nil, nil, 0)
	}
	comp := at.fn(value, n.value)
	if comp < 0 {
		n.left = at.insert(n.left, value)
	} else if comp > 0 {
		n.right = at.insert(n.right, value)
	}
	return balance(n)
}

func (at *AvlTree) remove(n *node, value interface{}) *node {
	if n == nil {
		return n
	}
	comp := at.fn(value, n.value)
	if comp < 0 {
		n.left = at.remove(n.left, value)
	} else if comp > 0 {
		n.right = at.remove(n.right, value)
	} else if n.left != nil && n.right != nil {
		min := at.findMin(n.right)
		n.right = at.remove(n.right, min)
		n.value = min
	} else {
		if n.left != nil {
			n = n.left
		} else {
			n = n.right
		}
	}
	return balance(n)
}

func (at *AvlTree) inOrder(n *node) {
	if n == nil {
		return
	}

	if n.left != nil {
		at.inOrder(n.left)
	}
	fmt.Print(n.value, " ")
	if n.right != nil {
		at.inOrder(n.right)
	}
}

func (at *AvlTree) findMin(n *node) interface{} {
	for n.left != nil {
		n = n.left
	}
	return n.value
}

// 平衡因子函数，调整平衡
func balance(n *node) *node {
	if n == nil {
		return nil
	}
	// 如果左子树大于右子树,说明插入的节点是在左子树
	if (maxH(n.left) - maxH(n.right)) > 1 {
		// 如果左子树的左子树高度大于左子树的右子树高度，单旋转
		if maxH(n.left.left) >= maxH(n.left.right) {
			n = rotateWithLeft(n)
		} else {
			n = doubleWithLeft(n)
		}
	} else if (maxH(n.right) - maxH(n.left)) > 1 { // 插入的节点在右子树
		if maxH(n.right.right) >= maxH(n.right.left) {
			n = rotateWithRight(n)
		} else {
			n = doubleWithRight(n)
		}
	}
	n.h = max(maxH(n.left), maxH(n.right)) + 1
	return n
}

// 左单旋
func rotateWithLeft(k2 *node) *node {
	k1 := k2.left
	k2.left = k1.right
	k1.right = k2
	k2.h = max(maxH(k2.left), maxH(k2.right)) + 1
	k1.h = max(maxH(k1.right), k2.h) + 1
	return k1
}

// 右单旋
func rotateWithRight(k1 *node) *node {
	k2 := k1.right
	k1.right = k2.left
	k2.left = k1
	k1.h = max(maxH(k1.left), maxH(k1.right)) + 1
	k2.h = max(maxH(k2.right), k1.h) + 1
	return k2
}

// 左双旋
func doubleWithLeft(k3 *node) *node {
	k3.left = rotateWithRight(k3.left)
	return rotateWithLeft(k3)
}

// 右双旋
func doubleWithRight(k2 *node) *node {
	k2.right = rotateWithLeft(k2.right)
	return rotateWithRight(k2)
}

func max(n1, n2 int) int {
	return int(math.Max(float64(n1), float64(n2)))
}

// 获取最大高度
func maxH(n *node) int {
	if n == nil {
		return -1
	} else {
		return n.h
	}
}
